### DISCUSSION TITLE
Roadmap for Logging

### DISCUSSION THREAD

**Discussion by MrMoose** [googleapis/google-cloud-cpp#6272] [General]
Hello everyone,

I am making good progress integrating the Google Cloud C++ SDK into my project. Thanks to your help here!

What I am curious about are your plans with regards to logging. This is one big thing that is missing in my integration. From the state of the source it appears to me as if logging support is planned but not yet available. I would like to enquire if you can make any statement about when one could expect this to be ready. 

Please note that basic functionality is sufficient. I don't need very much in terms of managing, deleting or retrieving logs. Only means to create a log group and log stream and push log entries via an SDK call.

Do you have any kind of plans or roadmap that may give me an idea of when I can expect this?

Cheers,

Moose

**Comment by devjgm** [comment]
Hi, Moose.

I'm very glad to hear that you're making progress! We're always happy to help.

Sadly, I cannot comment about when the Logging library will be GA (Generally Available). But I will note that you're not the only customer asking for logging. So, we know that people want it :wink: As you noticed, we do have some not-yet-production-ready logging code at https://github.com/googleapis/google-cloud-cpp/tree/master/google/cloud/logging. You're welcome to take a look at that it offer feedback as it whether or not it would meet your needs if/when it became GA. But do keep in mind that pre-GA code like that may still change in backward *incompatible* ways. Indeed, we know that https://github.com/googleapis/google-cloud-cpp/issues/6074 is going to be fixed soon and that will rename the files, which would likely break builds.

When there's more that I can share about logging, I'll post back here.

Take care.

**Reply by MrMoose** [reply]
Hello Greg,

thanks for your reply. At least I know you are aware of the need and that's something. I will have a look at the experimental code. Perhaps I can make use of it. About the possibility it breaks, well, I always expect this with new libs so no worries. As long as I can still get it to work I am quite all right.

Thanks a bunch and a great weekend to y'all!

Moose

**Reply by MrMoose** [reply]
> Hi, Moose.
> 
> I'm very glad to hear that you're making progress! We're always happy to help.

A little update on this.
I have revisited this question these days and decided to attempt a Logging integration based upon the classes available in 1.30.0, knowing that you don't consider this GA.
And it seems to work fine so far.
If you are interested, the implementation is here:
https://github.com/MrMoose/CloudConnector/blob/6007c65081cc689c5ddb3ca7b36a6c5b7e099c04/Source/CloudConnector/Private/GoogleLoggingOutputDevice.cpp#L153

I didn't encounter any issues apart from figuring out how the objects belong together and how to interpret the return value. Also having implemented the AWS logging before I missed the notion of that "continuation token" they have that you have to drag along between the calls to signify the order of the logs. However, not having this and going by the timestamp alone is much easier anyway.

So, just wanted to let you know that logging works fine for me now.

Cheers,
Moose

**Reply by devjgm** [reply]
I'm glad to hear it and thanks for letting us know! 👍🏼
